kale/embed/mogonet.py

```diff
@@ -0,0 +1,121 @@
+# =============================================================================
+
+"""
+Construct a message passing network using PyTorch Geometric for MOGONET method.
+
+This code is written by refactoring MOGONET model (https://github.com/txWang/MOGONET/blob/main/models.py) within
+'MessagePassing' base class provided in the PyTorch Geometric.
+
+Reference:
+Wang, T., Shao, W., Huang, Z., Tang, H., Zhang, J., Ding, Z., Huang, K. (2021). MOGONET integrates multi-omics data
+using graph convolutional networks allowing patient classification and biomarker identification. Nature communications.
+https://www.nature.com/articles/s41467-021-23774-w
+"""
+
+from typing import List, Optional, Union
+
+import torch
+import torch.nn.functional as F
+import torch_sparse
+from torch import Tensor
+from torch.nn import Module, Parameter
+from torch.nn.init import xavier_normal_
+from torch_geometric.nn.aggr import Aggregation
+from torch_geometric.nn.conv import MessagePassing
+from torch_sparse import SparseTensor
+
+
+class MogonetGCNConv(MessagePassing):
+    r"""Create message passing layers for MOGONET method.
+
+    Args:
+        in_channels (int): Size of each input sample.
+        out_channels (int): Size of each output sample.
+        bias (bool, optional): If set to :obj:`False`, the layer will not learn an additive bias. (default: :obj:`True`)
+        aggr (string or list or Aggregation, optional): The aggregation scheme to use,
+            *e.g.*, :obj:`"add"`, :obj:`"sum"` :obj:`"mean"`, :obj:`"min"`, :obj:`"max"` or :obj:`"mul"`.
+        **kwargs (optional): Additional arguments of :class:`torch_geometric.nn.conv.MessagePassing`.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        bias: bool = True,
+        aggr: Optional[Union[str, List[str], Aggregation]] = "add",
+        **kwargs,
+    ) -> None:
+        super().__init__(aggr=aggr, **kwargs)
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.weight = Parameter(torch.Tensor(self.in_channels, self.out_channels))
+
+        if bias:
+            self.bias = Parameter(torch.Tensor(self.out_channels))
+        else:
+            self.register_parameter("bias", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        """Initialize the parameters of the model."""
+        xavier_normal_(self.weight.data)
+
+        if self.bias is not None:
+            self.bias.data.fill_(0.0)
+
+    def forward(self, x: Tensor, edge_index: SparseTensor) -> Tensor:
+        # The format of edge_index is SparseTensor which performs fast sparse-matrix multiplication
+        x = torch.mm(x, self.weight)
+        out = self.propagate(edge_index, x=x)
+        return out
+
+    def message(self, x_j: Tensor) -> Tensor:
+        r"""Construct messages from node :math:`j` to node :math:`i` in analogy to :math:`\phi_{\mathbf{\Theta}}` for
```

Comment:
Can you use plain text for doc? If you like, supplement with math, but plain text explanation is important.

Reply:
The same for below